### PR TITLE
PBE run link to existing channel

### PR DESCRIPTION
**Pull request by trilopin**
## Summary
Add the ability to configure playbooks that can start runs linked to an existing channel.
This PR does not cover:
- multiple runs list in channels rhs
- ability to use different config than the one in PBE when starting a run

### Feature flag
There is a new feature flag:
* sever-side: `	LinkRunToExistingChannelEnabled bool` at Configuration
* webapp: `selectLinkRunToExistingChannelEnabled` selector and `useLinkRunToExistingChannelEnabled` hook available

The main idea is to merge to master (be able to deploy) as often as possible instead of working too much in integration branches

To enable it, in server config.json add `"linkruntoexistingchannelenabled": true,` to Playbooks config.

The feature flag is false (disabled) by default, so if it's not configured disabled will be assumed. It's expected to be set at true by the end of the V0. 

### API & GQL
The playbook query carries the new fields and `UpdatePlaybook` mutation accepts them.

### Service & Store
Two new fields in the playbook: `ChannelID` and `ChannelMode`.

The new fields have been created in the playbook, create-new-channel is the default behavior.
- (postgre) Removed the unique constraint for the channelID index in the ir_incident table.
- (MySQL) removed unique index (and kept the other index in channeled)

**Name field in ir_incident**
This field wasn't used and playbookrun store used channel.displayname instead.
Less than 2% of the runs had a non-empty name in the community server. That's why there's a migration to update all empty names by the matching channel display name.

### Playbook UI
The user can choose whether create a new channel per run or link an existing one. Available channels are those you have permission to.

Channel Selector has been modified to allow single channel selection (it was hardcoded top multi before)

It is considered valid to save without selecting a channel but telling that you'll link the run to an existing channel. It will be mandatory in the run-start.

https://user-images.githubusercontent.com/4096774/200348951-daf36d56-c9eb-409c-b3f4-f14a387eb256.mp4

**How it looks if feature flag is disabled (default)**
Screenshot 2022-11-10 at 16 54 26

## Channel selector
Already existing channel selector has been upgraded to:
- be usable for one channel (previously only multi)
- display options nicely: icon for private/public and some styles
Screenshot 2022-11-11 at 12 55 26
Screenshot 2022-11-11 at 12 54 58
Screenshot 2022-11-11 at 12 50 42

## Ticket Link
Fixes https://github.com/mattermost/mattermost-plugin-playbooks/issues/1541

## Checklist

- [ ] ~~Telemetry updated~~
- [X] Gated by an experimental feature flag
- [X] Unit tests updated

### CODE REVIEW

File: server/api/graphql_root_playbook.go [hunk 1]
```diff
@@ -122,6 +122,8 @@ func (r *PlaybookRootResolver) UpdatePlaybook(ctx context.Context, args struct {
 		IsFavorite                              *bool
 		CreateChannelMemberOnNewParticipant     *bool
 		RemoveChannelMemberOnRemovedParticipant *bool
+		ChannelID                               *string
```
Comment (crspeller) [inline server/api/graphql_root_playbook.go:125]: Just `ChannelID` confused me on first pass. Maybe call it `RunLinkChannelID`? Or something else more descriptive?

File: server/app/playbook_run_service.go [hunk 2]
```diff
@@ -253,30 +259,30 @@ func (s *PlaybookRunServiceImpl) CreatePlaybookRun(playbookRun *PlaybookRun, pb
...
 		playbookRun.ChannelID = channel.Id
 	} else {
-		existingPlaybookRunID, err := s.GetPlaybookRunIDForChannel(playbookRun.ChannelID)
-		if err != nil && !errors.Is(err, ErrNotFound) {
-			return nil, err
-		} else if existingPlaybookRunID != "" {
-			return nil, errors.Wrapf(ErrMalformedPlaybookRun, "playbook run %s already exists for channel %s", existingPlaybookRunID, playbookRun.ChannelID)
-		}
-
+		/*	PRE MULTIPLERUNS IN A CHANNEL
```
Comment (crspeller) [inline server/app/playbook_run_service.go:269]: We can just remove this.

File: server/sqlstore/playbook_run.go [hunk 3]
```diff
@@ -179,7 +179,7 @@ func NewPlaybookRunStore(pluginAPI PluginAPIClient, sqlStore *SQLStore) app.Play
 
 	// When adding a PlaybookRun column #1: add to this select
 	playbookRunSelect := sqlStore.builder.
-		Select("i.ID", "c.DisplayName AS Name", "i.Description AS Summary", "i.CommanderUserID AS OwnerUserID", "i.TeamID", "i.ChannelID",
+		Select("i.ID", "i.Name AS Name", "i.Description AS Summary", "i.CommanderUserID AS OwnerUserID", "i.TeamID", "i.ChannelID",
```
Comment (crspeller) [inline server/sqlstore/playbook_run.go:182]: We can remove the join on the channels table right?

**Review by crspeller — CHANGES_REQUESTED** [review]
Looking good!

**Review by gbochora — APPROVED** [review]
Great work, @trilopin 🥳

File: server/sqlstore/playbook_run.go (hunk 3 above)
Comment (trilopin) [reply server/sqlstore/playbook_run.go:182]: I also created a ticket for the refactor of the digest queries  https://github.com/mattermost/mattermost-plugin-playbooks/issues/1571

File: server/api/graphql_root_playbook.go (hunk 1 above)
Comment (trilopin) [reply server/api/graphql_root_playbook.go:125]: 2/5 if the playbook acts as a run template, wouldn't the same naming (for the fields that are copied) be the less surprising choice?

Comment (crspeller) [reply server/api/graphql_root_playbook.go:125]: That's a good point. I can see it both ways. I will leave it to your discretion.

File: server/app/playbook_run_service.go [hunk 4]
```diff
@@ -253,30 +259,30 @@ func (s *PlaybookRunServiceImpl) CreatePlaybookRun(playbookRun *PlaybookRun, pb
...
 		channel, err = s.pluginAPI.Channel.Get(playbookRun.ChannelID)
 		if err != nil {
 			return nil, err
 		}
 
-		playbookRun.Name = channel.Name
+	}
+
+	if playbookRun.Name == "" {
+		playbookRun.Name = pb.ChannelNameTemplate
```
[the remark this answers is not in the document]
Comment (trilopin) [reply server/app/playbook_run_service.go:280]: not really

File: server/sqlstore/migrations/future/postgres/0.61.down.sql [hunk 5]
```diff
@@ -0,0 +1,11 @@
+ALTER TABLE IR_Playbook DROP COLUMN IF EXISTS ChannelID;
+ALTER TABLE IR_Playbook DROP COLUMN IF EXISTS ChannelMode;
+
+-- Drop unique constraint and kee the index
+ALTER TABLE IR_Incident  ADD CONSTRAINT ir_incident_channelid_key UNIQUE(ChannelID);
```
Comment (trilopin) [inline server/sqlstore/migrations/future/postgres/0.61.down.sql:5]: I tried IF NOT EXISTS but apparently does not work here. thoughts @gbochora ?

**Review by abhijit-singh — APPROVED** [review]
Looks good!

File: server/sqlstore/migrations/future/postgres/0.61.down.sql (hunk 5 above)
Comment (gbochora) [reply server/sqlstore/migrations/future/postgres/0.61.down.sql:5]:
```suggestion
DO
$$
BEGIN
    IF NOT EXISTS (
        SELECT INDEXNAME FROM PG_INDEXES
        WHERE TABLENAME = 'ir_incident'
        AND INDEXNAME = 'ir_incident_channelid_key'
    ) THEN
        ALTER TABLE IR_Incident ADD CONSTRAINT ir_incident_channelid_key UNIQUE(ChannelID);
	END IF;
END
$$;

```

**Review by gbochora — COMMENTED** [review]
I've left a few comments regarding morph migrations.

File: server/sqlstore/migrations/future/mysql/0.61.down.sql [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    'ALTER TABLE IR_Playbook DROP COLUMN ChannelMode;',
+    'SELECT 1;'
+));
+
+PREPARE dropColumnIfExists FROM @preparedStatement;
+EXECUTE dropColumnIfExists;
+DEALLOCATE PREPARE dropColumnIfExists;
+
+SET @preparedStatement = (SELECT IF(
+    EXISTS(
```
[the remark this answers is not in the document]
Comment (trilopin) [reply server/sqlstore/migrations/future/mysql/0.61.down.sql:32]: oops

**Review by crspeller — APPROVED** [review]

**Pull request merged**